spring/src/main/java/io/cucumber/spring/SpringFactory.java

```diff
@@ -249,8 +251,10 @@ private ConfigurableApplicationContext getContext() {
 
         private void registerGlueCodeScope(ConfigurableApplicationContext context) {
             do {
-                context.getBeanFactory().registerScope(SCOPE_CUCUMBER_GLUE, new GlueCodeScope());
-                context = (ConfigurableApplicationContext) context.getParent();
+                synchronized (context) {
+                    context.getBeanFactory().registerScope(SCOPE_CUCUMBER_GLUE, new GlueCodeScope());
```

Comment:
registerScope is NOT threadsafe.  

In AbstractBeanFactory in Spring:

Scope previous = (Scope)this.scopes.put(scopeName, scope); //scopes is a LinkedHashMap